bin/liquidator/src/rebalance.rs

```diff
@@ -423,6 +409,301 @@ impl Rebalancer {
+                token.name
+            ))?;
+
+        let side = if native_amount < 0 {
+            Serum3Side::Bid
+        } else {
+            Serum3Side::Ask
+        };
+
+        let distance_from_oracle_price_bp =
+            I80F48::from_num(self.config.limit_order_distance_from_oracle_price_bps)
+                * match side {
+                    Serum3Side::Bid => 1,
+                    Serum3Side::Ask => -1,
+                };
+        let price_adjustment_factor =
+            (I80F48::from_num(10_000) + distance_from_oracle_price_bp) / I80F48::from_num(10_000);
+
+        let limit_price =
+            (token_price * price_adjustment_factor * I80F48::from_num(market.coin_lot_size))
+                .to_num::<u64>()
+                / market.pc_lot_size;
+        let mut max_base_lots =
+            (native_amount.abs() / I80F48::from_num(market.coin_lot_size)).to_num::<u64>();
+
+        debug!(
+            side = match side {
+                Serum3Side::Bid => "Buy",
+                Serum3Side::Ask => "Sell",
+            },
+            token = token.name,
+            oracle_price = token_price.to_num::<f64>(),
+            price_adjustment_factor = price_adjustment_factor.to_num::<f64>(),
+            coin_lot_size = market.coin_lot_size,
+            pc_lot_size = market.pc_lot_size,
+            limit_price = limit_price,
+            native_amount = native_amount.to_num::<f64>(),
+            max_base_lots = max_base_lots,
+            "building order for rebalancing"
+        );
+
+        // Try to buy enough to close the borrow
+        if max_base_lots == 0 && native_amount < 0 {
+            info!(
+                "Buying a whole lot for token {} to cover borrow of {}",
+                token.name, native_amount
+            );
+            max_base_lots = 1;
+        }
+
+        if max_base_lots == 0 {
+            warn!("Could not rebalance token '{}' (native_amount={}) using limit order, below base lot size", token.name, native_amount);
+            return Ok(());
+        }
+
+        let mut account = account.clone();
+        let create_or_replace_ixs = self
+            .mango_client
+            .serum3_create_or_replace_account_instruction(&mut account, *market_index, side)
+            .await?;
+        let cancel_ixs =
+            self.mango_client
+                .serum3_cancel_all_orders_instruction(&account, *market_index, 10)?;
+        let place_order_ixs = self
+            .mango_client
+            .serum3_place_order_instruction(
+                &account,
+                *market_index,
+                side,
+                limit_price,
+                max_base_lots,
+                ((limit_price * max_base_lots * market.pc_lot_size) as f64 * 1.01) as u64,
+                Serum3SelfTradeBehavior::CancelProvide,
+                if self.config.limit_order_distance_from_oracle_price_bps == 0 {
+                    Serum3OrderType::PostOnly
```

Comment:
can you explain?

Reply:
I wanted to make the liquidator only be on the maker side when configured to execute at oracle price, but it would be a hidden feature and could lead to unexpected behavior (no exec even if price would match). I'll revert that.